### PR TITLE
pcsx2: darwin support, add maintainer

### PR DESCRIPTION
**Pull request by matteo-pacini**
## Description of changes

- Split the package in `linux.nix` and `darwin.nix`
- `darwin.nix` fetches the pre-compiled binaries from GitHub
    - Tried to compile this but fails on `aarch64-darwin`, requires a `x86_64-darwin` machine
    -  Cannot be compiled also as Xcode is needed to build Metal shaders
    - The pre-compiled app runs fine under Rosetta, the version is in sync with the Linux one
- The three files have been formatted with `nixfmt-rfc-style`.
- Added myself as maintainer for Darwin

Screenshot 2024-06-15 at 20 09 40

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by afh** [comment]
Result of `nixpkgs-review pr 320122` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pcsx2</li>
  </ul>
</details>

File: pkgs/by-name/pc/pcsx2/darwin.nix
Comment (afh) [inline pkgs/by-name/pc/pcsx2/darwin.nix:1]: ❓ What are your thoughts around adding support to build this package on `x86_64-darwin` instead of installing the pre-built binary from upstream?

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (afh) [inline pkgs/by-name/pc/pcsx2/package.nix:32]: Nice! 👍 

My first thought was to overwrite add `sourceProvenance` in the platforms specific package files, e.g. for `darwin.nix`:
```
meta = meta // {
  sourceProvenance = lib.sourceTypes.binaryNativeCode;
};
```

Not sure if there is any benefit to that, but thought I'd share…

**Review by afh — APPROVED** [review]
Nicely done, @matteo-pacini, please find below some hopefully helpful comments for potential future improvement…

**Comment by afh** [comment]
The app requires Rosetta to run, how could this dependency be declared in nixpkgs in general to differentiate between packages that run natively on Apple Silicon and packages needing Rosetta?

File: pkgs/by-name/pc/pcsx2/darwin.nix
Comment (matteo-pacini) [reply pkgs/by-name/pc/pcsx2/darwin.nix:1]: @afh Imho,  It would make this package harder to maintain - I would make it compile from source only if it's supported on both `aarch64-darwin` and `x86_64-darwin`.

I also expect `x86_64-darwin` to slowly lose traction and disappear, so any effort towards that architecture could potentially become a waste of maintainers' time.

This is just my personal opinion, and up to debate: I expect plenty of new Nix adopter to be on `aarch64-darwin`, rather than `x86_64-darwin`.

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (matteo-pacini) [reply pkgs/by-name/pc/pcsx2/package.nix:32]: I've seen plenty of different approaches for this, not sure which is the best / preferred way of doing it.

Happy to change it if you think it can be improved.

**Comment by matteo-pacini** [comment]
@afh That's a good point - my gut tells me this should only be declared as a `x86_64-darwin`, and then `aarch64-darwin` can download this anyway by using Nix config / extra platforms.

@NixOS/darwin-maintainers thoughts on this?

**Comment by SuperSamus** [comment]
FYI, upstream is working on natively supporting ARM.
`https://github.com/PCSX2/pcsx2/pull/11364`

**Comment by emilazy** [comment]
> That's a good point - my gut tells me this should only be declared as a `x86_64-darwin` derivation, and then `aarch64-darwin` can download this anyway by using Nix config / extra platforms.

I agree. There’s nothing about this app that is `aarch64-darwin` (currently); it just so happens that all `aarch64-darwin` machines support software that lets them run `x86_64-darwin` software.

File: pkgs/by-name/pc/pcsx2/darwin.nix
Comment (emilazy) [reply pkgs/by-name/pc/pcsx2/darwin.nix:1]: It might be worth reconsidering this in light of the upcoming Apple Silicon support?

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (afh) [reply pkgs/by-name/pc/pcsx2/package.nix:32]: I don't think it would be an improvement per se, just a different way of doing it. Until there is a good reason to change it or a better way to do it, I'd leave it as is 🙂

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (afh) [inline pkgs/by-name/pc/pcsx2/package.nix:29]: Given the latest comments as this PR stands now I kindly request the following change:
```suggestion
    platforms = [ "x86_64-linux" "x86_64-darwin" ];
```

File: pkgs/by-name/pc/pcsx2/darwin.nix
Comment (afh) [reply pkgs/by-name/pc/pcsx2/darwin.nix:1]: That makes sense, @matteo-pacini, I'm alright with this being `x86_64-darwin` until Apple Silicon supports lands upstream and this can be built from source.

**Comment by matteo-pacini** [comment]
@SuperSamus @afh @emilazy Thanks for the feedback, I'm going to change the derivation and make it natively compile for `x86_64-darwin`, so it's ready for when `aarch64-darwin` support will be officially released.

**Review by afh — CHANGES_REQUESTED** [review]
Thanks everyone for chiming in on this, very helpful 🙏

@matteo-pacini please find below updated comments 🙂

**Comment by matteo-pacini** [comment]
@emilazy hit a roadblock, got most of the project to compile, but it fails when trying to compile Metal shaders, as that would require the host's `xcrun` and `Xcode`.

What's the best way forward here?
1) Using tools on the host would be tricky, i.e. it will depend on Xcode - lot of impurity.
2) I could patch the app so the shaders are not compiled, just copied over, but then I would have to change the renderer pipeline, so those are compiled at runtime instead. Not impossible but this would be highly time consuming. Also we don't know for how long this will work and might require constant maintenance over time.
3) Not a big fan of this idea: fetch the binary dist from Github regardless, and copy over the pre-compiled `*.metallib` files to the final app bundle
4) We could just stick to the binary distribution as a temporary solution, and think how to bring Metal shader compiling into the packaging pipeline, so other derivations can benefit from it too, and revisit this at a later stage.

Pinging also @NixOS/darwin-maintainers, any input will be greatly appreciated.

**Comment by matteo-pacini** [comment]
@emilazy @afh reverted this PR to its previous status.

As stated in https://github.com/NixOS/nixpkgs/pull/320122#issuecomment-2185396923, some thinking is required to bring Metal shaders into the pipeline.
I don't thinks there's an immediate / obvious solution to this, so I suggest to go with option (4) from the previous comment, rather than having no Darwin support at all.

I'll bring this topic up on Matrix when I got some time 👍🏻

**Review by afh — APPROVED** [review]
I agree with you, @matteo-pacini.

Would you be interested in setting up a draft PR for the native darwin build that also tracks the development of Apple silicon in upstream? Might be helpful to get folks together and advise on best paths forward in regards to xcrun and Xcode…

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (toonn) [reply pkgs/by-name/pc/pcsx2/package.nix:32]: Don't really have an opinion on where to put the attribute, whatever works is fine.

However, packages that are pretty much just fetching a binary usually set `hydraPlatforms = []` because binaries don't really add value to the cache. In this case that would have to be `hydraPlatforms = lib.platforms.linux` or a different list as long as it excludes the Darwins.

Comment (matteo-pacini) [reply pkgs/by-name/pc/pcsx2/package.nix:33]: @toonn done, and thanks for the feedback!

**Review by afh — APPROVED** [review]
🙂👍

**Review by toonn — APPROVED** [review]
LGTM

**Pull request merged**
